src/wavm.cpp

```diff
@@ -0,0 +1,263 @@
+#include "wavm.h"
+#include "debugging.h"
+#include "eei.h"
+#include "exceptions.h"
+
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+
+using namespace std;
+
+namespace hera {
+
+
+std::unique_ptr<WasmEngine> WavmEngine::create()
+{
+  return std::unique_ptr<WasmEngine>{new WavmEngine};
+}
+
+
+namespace wavm_host_module {
+
+  // first a pointer to the ethereum interface, which is used in host functions
+  WavmEthereumInterface* interface;
+
+
+  // the host module is called 'ethereum'
+  DEFINE_INTRINSIC_MODULE(ethereum)
+
+
+  /*
+  finish
+    Set the returning output data for the execution. This will cause a trap and the execution will be aborted immediately.
+  Parameters
+    dataOffset i32ptr the memory offset of the output data (bytes)
+    length i32 the length of the output data
+  Returns
+    doesn't return
+  */
+  // some global variables, maybe these should go elsewhere
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "finish", void, finish, U32 dataOffset, U32 length)
+  {
+    printf("finish  dataOffset=%i  length=%i\n",dataOffset,length);
+    //dataOffset = static_cast< uint32_t > (dataOffset);
+    //dataOffset = static_cast< uint32_t > (length);
+
+    try{
+      interface->eeiFinish(dataOffset, length);
+    }
+    catch(hera::HeraException e){
+      //gracefully return from the wavm invokation
+      throwException(Runtime::Exception::calledAbortType);
+    }
+  }
+
+
+  /*
+  revert
+    Set the returning output data for the execution. This will cause a trap and the execution will be aborted immediately.
+  Parameters
+    dataOffset i32ptr the memory offset of the output data (bytes)
+    length i32 the length of the output data
+  Returns
+    doesn't return
+  */
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "revert", void, revert, U32 dataOffset, U32 length)
+  {
+    printf("revert  dataOffset=%i  length=%i\n",dataOffset,length);
+    try{
+      interface->eeiRevert(dataOffset, length);
+    }
+    catch(hera::HeraException e){
+      //gracefully return from the wavm invokation
+      throwException(Runtime::Exception::calledAbortType);
```

Comment:
We need a way to carry over the hera exception type.

Reply:
Yes, good point. I will look for a way to carry over the hera exception type.